File: JS-programs/E1/main.js
import WiFi from "wifi";
import Analog from "pins/analog";
import { Request } from "http";

setWakeUpTime();

new WiFi(
  {
    ssid: "xxx",
    password: "xxx",
  },
  function (status) {
    if (status == WiFi.gotIP) {
      // network is ready  
      const value = readTemp();
      sendData( value)
    }
  }
);

function sendData(temp) {
  let request = new Request({
    host: "api.thingspeak.com",
    path: `/update?api_key=<HIDE>=${temp}`,
    response: String,
  });
  request.callback = function (message, value, etc) {
    if (Request.responseComplete == message) {
      // go to sleep
      sleep();
    }
  };
}

function readTemp() {
  const readedValue = Analog.read(0);
  return ((readedValue / 1024.0) * 3300) / 10;
}

// native interface to ESP-IDF (see main.c file)
function setWakeUpTime() @ "xs_setWakeUpTime";
function sleep() @ "xs_sleep";

File: JS-programs/E1/main.c
// call to native ESP-IDK tolls to wake up and sleep
#include "xsmc.h"

void xs_setWakeUpTime(xsMachine *the)
{
    uint64_t sleeptime = UINT64_C(15 * 1000000);
    esp_sleep_enable_timer_wakeup(sleeptime);
}
void xs_sleep(xsMachine *the)
{
    esp_deep_sleep_start();  
}